forms/user/registration_form_test.go

```diff
@@ -0,0 +1,210 @@
+				}
+
+				formValidation := validation.Validation{}
+				form.Valid(&formValidation)
+
+				Expect(len(formValidation.Errors)).To(BeZero())
+			})
+		})
+
+		Context("given INVALID params", func() {
+			Context("given an existing email", func() {
+				It("produces an error", func() {
+					_ = FabricateUser("dev@nimblehq.co", "password")
+					form := form.RegistrationForm{
+						Email:           "dev@nimblehq.co",
+						Password:        "password",
+						ConfirmPassword: "password",
+					}
+
+					formValidation := validation.Validation{}
+					form.Valid(&formValidation)
+
+					Expect(len(formValidation.Errors)).To(Equal(1))
+					Expect(formValidation.Errors[0].Key).To(Equal("Email"))
+					Expect(formValidation.Errors[0].Message).To(Equal("Email is already in use"))
+				})
+			})
+
+			Context("given a mismatch confirm password", func() {
+				It("produces an error", func() {
+					form := form.RegistrationForm{
+						Email:           "dev@nimblehq.co",
+						Password:        "password",
+						ConfirmPassword: "pas____d",
+					}
+
+					formValidation := validation.Validation{}
+					form.Valid(&formValidation)
+
+					Expect(len(formValidation.Errors)).To(Equal(1))
+					Expect(formValidation.Errors[0].Key).To(Equal("ConfirmPassword"))
+					Expect(formValidation.Errors[0].Message).To(Equal("Confirm password confirmation must match"))
+				})
+			})
+		})
+	})
+
+	Describe("#Create", func() {
+		Context("given valid params", func() {
+			It("returns user object", func() {
+				form := form.RegistrationForm{
+					Email:           "dev@nimblehq.co",
+					Password:        "password",
+					ConfirmPassword: "password",
+				}
+
+				user, errors := form.Create()
+				if len(errors) > 0 {
+					Fail("Save a new user with registration form failed")
+				}
+
+				Expect(user).NotTo(BeNil())
+			})
+
+			It("does NOT produce any errors", func() {
+				form := form.RegistrationForm{
+					Email:           "dev@nimblehq.co",
+					Password:        "password",
+					ConfirmPassword: "password",
+				}
+
+				_, errors := form.Create()
+
+				Expect(len(errors)).To(BeZero())
+			})
+		})
+
+		Context("given INVALID params", func() {
+			Context("given an existing email", func() {
+				It("does NOT return a user object", func() {
+					_ = FabricateUser("dev@nimblehq.co", "password")
+					form := form.RegistrationForm{
+						Email:           "dev@nimblehq.co",
+						Password:        "password",
+						ConfirmPassword: "password",
+					}
+
+					user, _ := form.Create()
+
+					Expect(user).To(BeNil())
+				})
+
+				It("produces an error", func() {
+					_ = FabricateUser("dev@nimblehq.co", "password")
+					form := form.RegistrationForm{
+						Email:           "dev@nimblehq.co",
+						Password:        "password",
+						ConfirmPassword: "password",
+					}
+
+					_, errors := form.Create()
+
+					Expect(errors[0].Error()).To(Equal("Email is already in use"))
+				})
+			})
+
+			Context("given an INVALID email", func() {
```

Comment:
@Lahphim Same as the comment on the authentication form, can we add the cases when the params are missing?

Reply:
https://github.com/Lahphim/go-crawler-challenge/pull/46/commits/e5246f348ea4063d1b107866c34d33574ff7d3be